R/locks.R

```diff
@@ -0,0 +1,185 @@
+#'   * `add_table_lock()` adds a record in the schema.locks table with the current time and R-session process id.
+#'   * `remove_table_lock()` removes records in the schema.locks table with the target table and the
+#'      R-session process id.
+#'   * `is_lock_owner()` returns `TRUE` if the current process id (pid) matches the pid associated with the lock on
+#'      db_table in schema.locks. If no lock is found, `NULL` is returned.
+#' @template conn
+#' @param db_table (`character(1)`)\cr
+#'   A specification of "schema.table" to modify lock for.
+#' @param schema (`character(1)`)\cr
+#'   The schema where the "locks" table should be created.
+#' @return
+#'   Most have return value (called for side effects).
+#'   `is_lock_owner()` returns the TRUE if the process can modify the table.
+#' @examples
+#'   conn <- DBI::dbConnect(RSQLite::SQLite())
+#'
+#'   is_lock_owner(conn, "test_table") # NULL
+#'
+#'   add_table_lock(conn, "test_table")
+#'   is_lock_owner(conn, "test_table") # TRUE
+#'
+#'   remove_table_lock(conn, "test_table")
+#'   is_lock_owner(conn, "test_table") # NULL
+#'
+#'   DBI::dbDisconnect(conn)
+#' @export
+add_table_lock <- function(conn, db_table, schema = NULL) {
+  checkmate::assert_class(conn, "DBIConnection")
+  assert_id_like(db_table)
+  checkmate::assert_character(schema, null.ok = TRUE)
+
+  # Determine lock table id
+  lock_table_id <- id(paste(c(schema, "locks"), collapse = "."), conn)
+
+  # Create lock table if missing
+  if (!table_exists(conn, lock_table_id)) {
+    suppressMessages(
+      dplyr::copy_to(
+        conn,
+        data.frame(
+          "schema" = character(0),
+          "table" = character(0),
+          "lock_start" = numeric(0),
+          "pid" = numeric(0)
+        ),
+        lock_table_id,
+        temporary = FALSE
+      )
+    )
+
+    if (inherits(conn, "PqConnection")) { # PostgreSQL needs an index for rows_insert
+      res <- DBI::dbSendQuery(conn, glue::glue("ALTER TABLE {lock_table_id} ADD PRIMARY KEY (\"schema\", \"table\");"))
+      DBI::dbClearResult(res)
+    }
+  }
+
+  # Get a reference to the table
+  lock_table <- dplyr::tbl(conn, lock_table_id, check_from = FALSE)
+
+  # We then try to insert a lock, if none exists, our process ID (pid) will be assigned to the table-
+  # If one already exists, our insert will fail.
+  tryCatch(
+    {
+      db_table_id <- id(db_table, conn)
+
+      lock <- dplyr::copy_to(
+        conn,
+        data.frame(
+          "schema" = purrr::pluck(db_table_id, "name", "schema"),
+          "table" = purrr::pluck(db_table_id, "name", "table"),
+          "pid" = Sys.getpid(),
+          "lock_start" = as.numeric(Sys.time())
+        ),
+        name = unique_table_name()
+      )
+
+      dplyr::rows_insert(lock_table, lock, by = c("schema", "table"), conflict = "ignore", in_place = TRUE)
+    },
+    error = function(e) {
+      print(e$message)
+    }
+  )
+
+  return(NULL)
```

Comment:
Suggestion: Rather than having to check for ownership afterwards then one could return TRUE if a lock was added and FALSE if the table was already locked.